### PR TITLE
feat: improve the solana memo program demo by adding a counter

### PR DESCRIPTION
**Pull request by roberts-ivanovs**
**Referenced issue**

Closes https://github.com/eigerco/solana-axelar/issues/256

**Summary of changes**

The previous "demo" between EVM and Solana memo programs in a sense was superficial because it did not showcase how mutable state works for cross-chain calls.
This PR improves that by adding a "Counter" PDA that's owned by the memo program, and the counter gets incremented when a new memo gets received from the Solana Gateway.

- [x] Does this PR change behavior or adds a new feature ?
- [x] docs updated
- [x] tests updated
- [x] code style matches guidelines

**Reviewer recommendations**

How to make the review easier. i.e "Its a big PR and the recommendation is to go through each commit" or "In order to run the tests, execute ..."

### CODE REVIEW

File: solana/helpers/test-fixtures/src/test_setup.rs [hunk 1]
```diff
@@ -734,6 +734,44 @@ impl TestFixture {
...
+        gatewa_execute_data: &gateway::state::GatewayExecuteData,
+        decoded_payload: &DataPayload<'a>,
+        gateway_approved_command_pdas: &[solana_sdk::pubkey::Pubkey],
+        gateway_root_pda: solana_sdk::pubkey::Pubkey,
+    ) -> solana_program_test::BanksTransactionResultWithMetadata {
+        let DecodedCommand::ApproveContractCall(approved_message) =
+            gatewa_execute_data.command_batch.commands[0].clone()
+        else {
+            panic!("expected ApproveContractCall command")
+        };
```
Comment (roberts-ivanovs) [inline solana/helpers/test-fixtures/src/test_setup.rs:763]: this needs to be removed

File: solana/programs/axelar-solana-memo-program/src/instruction.rs [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+
+/// Instructions supported by the Axelar Memo program.
+#[repr(u8)]
+#[derive(Clone, Debug, PartialEq, BorshSerialize, BorshDeserialize)]
+pub enum AxelarMemoInstruction {
+    /// Initialize the memo program by creating a counter PDA
+    ///
+    /// Accounts expected by this instruction:
+    ///
+    /// 0. [] payer
```
Comment (roberts-ivanovs) [inline solana/programs/axelar-solana-memo-program/src/instruction.rs:19]: update comment to set the `payer` as `signer`

File: solana/programs/axelar-solana-memo-program/src/instruction.rs [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+        AccountMeta::new_readonly(gateway::ID, false),
+    ];
+    Ok(Instruction {
+        program_id: crate::ID,
+        accounts,
+        data: to_vec(&instruction_data)?,
+    })
+}
+
+/// Helper function to build a memo payload instruction (to simulate )
```
Comment (roberts-ivanovs) [inline solana/programs/axelar-solana-memo-program/src/instruction.rs:98]: finish the "end" of the comment

File: solana/programs/axelar-solana-memo-program/src/processor.rs [hunk 4]
```diff
@@ -102,7 +110,43 @@ pub fn process_native_ix(
...
+
+    let counter = crate::state::Counter { counter: 0, bump };
+
+    // Check: System Program Account
+    if !system_program::check_id(system_account.key) {
+        return Err(ProgramError::IncorrectProgramId);
+    }
+    // Check: Memo counter PDA Account is not initialized
+    counter_pda.check_uninitialized_pda()?;
+    // // Check: counter PDA account uses the canonical bump.
```
Comment (roberts-ivanovs) [inline solana/programs/axelar-solana-memo-program/src/processor.rs:141]: remove  `//`

**Review by eloylp — COMMENTED** [review]
Hey @roberts-ivanovs , initially the code LGTM, i am not approving this just because there are some commits (at the beginning) identical to the already merged #245 . Is this because the branch wasn't rebased with main ? Well, sorry its a bit late for me now and maybe i do not see things clear right now. Just let me know and i will approve it.

image

File: solana/programs/axelar-solana-memo-program/src/processor.rs [hunk 5]
```diff
@@ -102,7 +110,43 @@ pub fn process_native_ix(
...
+        AxelarMemoInstruction::Initialize { counter_pda_bump } => {
+            process_initialize_memo_prograrm_counter(program_id, accounts, counter_pda_bump)?;
+        }
     }
 
     Ok(())
 }
+
+/// This function is used to initialize the program.
+pub fn process_initialize_memo_prograrm_counter(
```
Comment (eloylp) [inline solana/programs/axelar-solana-memo-program/src/processor.rs:122]:
```suggestion
pub fn process_initialize_memo_program_counter(
```

File: solana/programs/axelar-solana-memo-program/src/instruction.rs (hunk 2 above)
Comment (roberts-ivanovs) [reply solana/programs/axelar-solana-memo-program/src/instruction.rs]: fixed in 5c1800368c20ab8593588c8daeb021eb0b2e9599

File: solana/programs/axelar-solana-memo-program/src/instruction.rs (hunk 3 above)
Comment (roberts-ivanovs) [reply solana/programs/axelar-solana-memo-program/src/instruction.rs]: fixed in 5c1800368c20ab8593588c8daeb021eb0b2e9599

File: solana/programs/axelar-solana-memo-program/src/processor.rs (hunk 4 above)
Comment (roberts-ivanovs) [reply solana/programs/axelar-solana-memo-program/src/processor.rs]: fixed in 5c1800368c20ab8593588c8daeb021eb0b2e9599

File: solana/programs/axelar-solana-memo-program/src/processor.rs [hunk 6]
```diff
@@ -102,7 +110,43 @@ pub fn process_native_ix(
                 &[sender.clone(), gateway_root_pda.clone()],
             )?;
         }
+        AxelarMemoInstruction::Initialize { counter_pda_bump } => {
+            process_initialize_memo_prograrm_counter(program_id, accounts, counter_pda_bump)?;
```
[the remark this answers is not in the document]
Comment (roberts-ivanovs) [reply solana/programs/axelar-solana-memo-program/src/processor.rs]: fixed in 5c1800368c20ab8593588c8daeb021eb0b2e9599

File: solana/programs/axelar-solana-memo-program/src/processor.rs (hunk 5 above)
Comment (roberts-ivanovs) [reply solana/programs/axelar-solana-memo-program/src/processor.rs]: fixed in 5c1800368c20ab8593588c8daeb021eb0b2e9599

File: solana/helpers/test-fixtures/src/test_setup.rs (hunk 1 above)
Comment (roberts-ivanovs) [reply solana/helpers/test-fixtures/src/test_setup.rs]: fixed in e91d847cdc7c26f29af080de6c2822ca61ff89a1

**Comment by roberts-ivanovs** [comment]
> Is this because the branch wasn't rebased with main ?

yes, you are right. I restructured the git history to make it easier to read 👍

**Comment by roberts-ivanovs** [comment]
And the CI situation has been resolved. Thanks for bringing this to light 🙏

**Review by eloylp — APPROVED** [review]

**Pull request merged**